src/site/content/en/blog/more-capable-form-controls/index.md

```diff
@@ -0,0 +1,306 @@
+
+### Input validation
+
+Your control can also participate in form validation by calling the `setValidity()`
+method on the internals object.
+
+```js
+// Assume this is called whenever the internal value is updated
+onUpdateValue() {
+  if (!this.matches(':disabled') && this.hasAttribute('required') &&
+      this._value < 0) {
+    this._internals.setValidity({customError: true}, 'Value cannot be negative.');
+  } 
+  else {
+    this._internals.setValidity({});
+  }
+  this.internals.setFormValue(this._value);
+}
+```
+
+You can style a form-associated custom element with the `:valid` and `:invalid`
+pseudoclasses, just like a built-in form control.
+
+{% Aside %}Although you can set a validation message, Chrome currently
+fails to display the validation message for form-associated custom elements.{% endAside %}
+
+### Lifecycle callbacks
+
+A form-associated custom element API includes a set of extra lifecycle callbacks to tie in to the form lifecycle. The callbacks are optional: only implement a callback if your element needs to do something at that point in the lifecycle.
+
+#### `void formAssociatedCallback(form)`
+
+Called when the browser associates the element with a form element, or disassociates the element from a form element.
+
+#### `void formDisabledCallback(disabled)`
+
+Called after the `disabled` state of the element changes, either because the `disabled` attribute of this element was added or removed; or because the `disabled` state changed on a `<fieldset>` that's an ancestor of this element. The `disabled` parameter represents the new [disabled state](https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#concept-fe-disabled) of the element. The element may, for example, disable elements in its shadow DOM when it is disabled.
+
+#### `void formResetCallback()`
+
+Called after the form is reset. The element should reset itself to some kind of default state. For native inputs, this usually involves setting the `value` property to match the `value` attribute set in markup (or in the case of a checkbox, setting the `checked` property to match the `checked` attribute.
+
+#### `void formStateRestoreCallback(state, mode)`
+
+Called in one of two circumstances:
+
+*   When the browser restores the state of the element (for example, after a navigation, or when the browser restarts).  The <code>mode</code> argument is <code>"restore"</code> in this case.
+*   When the browser's input-assist features such as form autofilling sets a value. The <code>mode</code> argument is <code>"autocomplete"</code> in this case.
+
+The type of the first argument depends on how the `setFormValue()` method was called. For more details, see [Restoring form state](#restoring-form-state).
+
+### Restoring form state {: #restoring-form-state }
+
+Under some circumstances—like when navigating back to a page, or restarting the browser, the brower may try to restore the state of a form to the state the user left it in.
```

Comment:
'brower' -> 'browser'

'restore the state of a form to the state...' -> 'restore the form to the state...'

Reply:
👍